File: trunk/src/python/pyasynchio/apoll.py
# Copyright (c) 2005 Vladimir Sukhoy

# Permission is hereby granted, free of charge, to any person obtaining a copy 
# of this software and associated documentation files (the "Software"), to deal 
# in the Software without restriction, including without limitation the rights 
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
# copies of the Software, and to permit persons to whom the Software is furnished 
# to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all 
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
# INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
# PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION 
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

__doc__="""Module which implements apoll class wrapper"""


from _pyasynchio import apoll as _apoll, get_sock_family, get_sock_type, get_sock_proto
import socket

class apoll(_apoll):
    """\
    The apoll class is responsible for both initiating asynchronous operations
    and delivering their results. The results are delivered via simple 
    interface which does not bring extra OOP burden for class user. Results
    are delivered as simple dictionaries and can be either processed by some
    object oriented logic or by more primitive approaches. That is why this
    class is called apoll, or "asynchronous poll". The approach taken is similar
    to the approach of poll object in module select. Nevertheless, AIO which
    apoll implements is very much different from nonblocking I/O which is 
    implemented in select module.
    """
    def __init__(self, max_threads = 0):
        """__init__([max_threads=0])
        Initialize apoll object. max_threads parameter specifies how
        many threads are able to poll this object simultaneously.
        """
        _apoll.__init__(self, max_threads)

    def accept(self, lsock, act = None):
        """accept(self, lsock [,act])
        Start asynchronous accept operation on listening socket. 
        act, which defaults to None is asynchronous completion token for this operation. 
        Upon completion, the result dictionary has following keys:
        "type" maps to string  "accept",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "listen_socket" maps to listening socket (lsock parameter of accept), 
        "accept_socket" maps to socket which is the result of accept op, 
        "act" maps to asynchronous completion token of this operation (act parameter of accept).
        """
        lsock_family = get_sock_family(lsock._sock)
        lsock_type = get_sock_type(lsock._sock)
        lsock_proto = get_sock_proto(lsock._sock)
        asock = socket.socket(lsock_family, lsock_type, lsock_proto)
        return _apoll.accept(self, lsock._sock, asock._sock, lsock, asock, act)

    def connect(self, sock, addr, act = None):
        """connect(self, sock, addr, [, act])
        Start asynchronous connect operation on socket sock.
        If sock is SOCK_STREAM socket, then if it is not bound, then it will be 
        bound to random address (just like synchronous connect does).
        act, which defaults to None represents asynchronous completion token for this operation.
        Upon completion the result dictionary has following keys:
        "type" maps to string "connect",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "socket" maps to socket which was connected (sock parameter of apoll.connect), 
        "addr" maps to addr object which was passed to this method (addr parameter of apoll.connect)
        "act" maps to asynchronous completion token for this operation (act parameter of apoll.connect).
        """
        if get_sock_type(sock._sock) == socket.SOCK_STREAM:
            try:
                sock.bind(('', 0))
            except socket.error:
                pass
        return _apoll.connect(self, sock._sock, sock, addr, act)

    def send(self, sock, data, flags = 0, act = None):
        """send(self, sock, data [, flags, act])
        Start asynchronous send operation on socket, flags have the same meaning as in usual socket.send method.
        act, which defaults to None represents asynchronous completion token for this operation.
        Upon completion the result notification dictionary has following keys:
        "type" maps to string "send", 
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "socket" maps to socket which was used in send operation (sock parameter of apoll.send method), 
        "data" maps to data which was intended to be sent (data parameter of apoll.send method), 
        "flags" maps to flags for send operation (flags parameter of apoll.send method),
        "count" maps to integer which specifies how many bytes of data was sent,
        "act" maps to asynchronous completion token for this operation (act parameter of apoll.send method).
        """
        return _apoll.send(self, sock._sock, sock, data, flags, act)

    def sendto(self, sock, addr, data, flags = 0, act = None):
        """sendto(self, sock, addr, data [, flags, act])
        Start asynchronous sendto operation on socket, flags have the same meaning as in synchronous socket.sendto method.
        act, which defaults to None represents asynchronous completion token for this operation.
        Upon completion the result notification dictionary has following keys:
        "type" maps to string "sendto",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "socket" maps to socket which was used in sendto operation (sock parameter of apoll.sendto method),
        "data" maps to data which was intended to be sent (data parameter of apoll.sendto method),
        "flags" maps to flags for this particular sendto operation (flags parameter of apoll.sendto),
        "count" maps to integer which specifies how many bytes of data was sent,
        "addr" maps to address to which data was intended to be sent (addr parameter of apoll.sendto),
        "act" maps to asynchronous completion token for this operation (act parameter of apoll.sendto)        
        """
        return _apoll.sendto(self, sock._sock, sock, addr, data, flags, act)

    def recv(self, sock, bufsize, flags = 0, act = None):
        """recv(self, sock, bufsize [, flags , act])
        Start asynchronous recv operation on socket.
        act, which defaults to None represents asynchronous completion token for this operation.
        Upon completion the result notification dictionary has following keys:
        "type" maps to string "recv",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "socket" maps to socket which was used in recv operation (sock parameter of apoll.recv method),
        "bufsize" maps to buffer size for data (bufsize parameter of apoll.recv),
        "flags" maps to flags for this particular recv operation (flags parameter of apoll.recv),
        "act" maps to asynchronous completion token for this operation (act parameter of apoll.recv).
        If recv is successful, then result notification dictionary has one more key:
        "data" which maps to data being received (of size <= bufsize).
        """
        return _apoll.recv(self, sock._sock, sock, bufsize, flags, act)

    def recvfrom(self, sock, bufsize, flags = 0, act = None):
        """recvfrom(self, sock, bufsize, [, flags, act])
        Start asynchronous recvfrom operation on socket.
        act, which defaults to None represents asynchronous completion token for this particular operation.
        Upon completion the result notification dictionary has following keys:
        "type" maps to string "recvfrom",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "socket" maps to socket which was used in recvfrom operation (sock parameter of apoll.recvfrom method),
        "bufsize" maps to buffer size for data (bufsize parameter of apoll.recvfrom),
        "flags" maps to flags for this particular recvfrom operation (flags parameter of apoll.recvfrom),
        "act" maps to asynchronous completion token for this operation (act parameter of apoll.recvfrom).
        If recvfrom is successful, then result notification dictionary has one more key:
        "data" which maps to data being received (of size <= bufsize).
        """
        return _apoll.recvfrom(self, sock._sock, sock, bufsize, flags, act)

    def poll(self, timeout = None):
        """poll(self [, timeout]) -> list_of_result_dictionaries
        Poll asynchronous results. timeout (in milliseconds) specifies how much time to wait for first 
        result. If omitted or negative, then waiting is infinite. Returns list of dictionaries with
        string keys (all of them have "type" key which maps to string which specifies the name of
        operation, like "send", "recv", "read", "write", etc. matching the name of appropriate method). Each
        of dictionaries also contains "act" key which maps to asynchronous completion token of particular
        operation.
        """
        if timeout == None or timeout < 0:
            timeout = -1    
        return _apoll.poll(self, timeout)
    
    def cancel(self, obj):
        """cancel(self, obj)
        Makes an attempt to cancel pending AIO operation on file or socket (obj). This method is
        unreliable, e.g. on windows it cancels only AIO operations which were issued by current thread.
        Exercise caution.
        """
        return _apoll.cancel(self, obj)

    def read(self, file, offset, bufsize, act = None):
        """read(self, file, offset, size [, act])
        Start asynchronous read operation on file.
        act, which defaults to None represents asynchronous completion token for this particular operation.
        Upon completion the result notification dictionary has following keys:
        "type" maps to string "read",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "file" maps to file object (file parameter of apoll.read),
        "offset" maps to offset of read (offset parameter of apoll.read),
        "bufsize" maps to buffer size, that is number of bytes to read (bufsize parameter of apoll.read),
        "data" maps to data which was read,
        "act" maps to asynchronous completion token for this operation.
        """
        return _apoll.read(self, file, offset, bufsize, act)

    def write(self, file, offset, data, act = None):
        """write(self, file, offset, data [, act])
        Start asynchronous write operation on file.
        act, which defaults to None represents asynchronous completion token for this particular operation.
        Upon completion the result notification dictionary has following keys:
        "type" maps to string "write",
        "success" maps to 1 if operation succeeded and 0 otherwise,
        "file" maps to file object (file parameter of apoll.write),
        "offset" maps to offset of write (offset parameter of apoll.write),
        "data" maps to data which had to be written,
        "count" maps to number of bytes that were written,
        "act" maps to asynchronous completion token for this operation.
        """
        return _apoll.write(self, file, offset, data, act)
        

File: trunk/src/cplusplus/pyasynchio/apoll.cpp
/*
Copyright (c) 2005 Vladimir Sukhoy

Permission is hereby granted, free of charge, to any person obtaining a copy 
of this software and associated documentation files (the "Software"), to deal 
in the Software without restriction, including without limitation the rights 
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
copies of the Software, and to permit persons to whom the Software is furnished 
to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all 
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION 
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/


#include <pyasynchio/apoll.hpp>
#include <pyasynchio/utils.hpp>
#include <pyasynchio/aop.hpp>
#include <pyasynchio/socketmodule_stuff.hpp>
#include <new>
#include <internal.h>
#include <winternl.h>
#include <mswsock.h>
#include <string>

namespace pyasynchio {

apoll::apoll(::PyObject *args, ::PyObject *kwargs)
: apoll_impl(args, kwargs)
{
}

apoll::~apoll()
{
}

bool apoll::accept(::PySocketSockObject *lsocko
                      , ::PySocketSockObject *asocko
                      , ::PyObject *lsock_ref
                      , ::PyObject *asock_ref
                      , ::PyObject *acto)
{
    aop_accept *aaop = new aop_accept(acto				// acto
        , lsocko                                        // lso
        , asocko                                        // aso
        , lsock_ref                                     // lso_ref
        , asock_ref                                     // aso_ref
		);
    bool ar = accept_impl(aaop);
    if (!ar) {
        delete aaop;
    }
    return ar;
}

bool apoll::recv(::PySocketSockObject *so
                    , ::PyObject *so_ref
                    , unsigned long bufsize
                    , unsigned long flags
                    , ::PyObject *acto)
{
    aop_recv * arop = new aop_recv(acto, so, so_ref, bufsize, flags);
	bool rr = recv_impl(arop);
	if (!rr) {
		delete arop;
	}
	return rr;
}

bool apoll::recvfrom(::PySocketSockObject * so
                        , ::PyObject *so_ref
                        , unsigned long bufsize
                        , unsigned long flags
                        , ::PyObject * acto)
{
    aop_recvfrom *arfop = new aop_recvfrom(acto, so, so_ref, bufsize, flags);
	bool rfr = recvfrom_impl(arfop);
	if (!rfr) {
		delete arfop;
	}
	return rfr;
}

bool apoll::sendto(::PySocketSockObject *so, ::PyObject *so_ref
                      , ::PyObject *addro
                      , ::PyObject *datao
                      , unsigned long flags
                      , ::PyObject *acto)
{
    aop_sendto *asynch_sendto_op = new aop_sendto(acto, so, so_ref, addro
												, flags, datao);
	bool sendto_result = sendto_impl(asynch_sendto_op);
	if (!sendto_result) {
		delete asynch_sendto_op;
	}
	return sendto_result;
}

bool apoll::send(::PySocketSockObject *so, ::PyObject *so_ref
                    , ::PyObject *datao, unsigned long flags
                    , ::PyObject *acto)
{

    aop_send * asynch_send_op = new aop_send(acto, so, so_ref, flags, datao);
	bool send_result = send_impl(asynch_send_op);
	if (!send_result) {
		delete asynch_send_op;
	}
	return send_result;
}

bool apoll::connect(::PySocketSockObject *so, ::PyObject *so_ref
                       , ::PyObject *addro, ::PyObject *acto)
{
    aop_connect * acop = new aop_connect(acto, so, so_ref, addro);
    bool cr = connect_impl(acop);
    if (!cr) {
        delete acop;
    }
    return cr;
}

bool apoll::cancel(PyObject *o)
{
    HANDLE h;
    if(PyObject_IsInstance(o, reinterpret_cast<PyObject*>(socketmodule_api.Sock_Type))) {
        PySocketSockObject * so = reinterpret_cast<PySocketSockObject*>(o);
        h = reinterpret_cast<HANDLE>(so->sock_fd);
    }
    else {
        if (PyObject_IsInstance(o, reinterpret_cast<PyObject*>(&PyFile_Type))) {
            PyFileObject *fo = reinterpret_cast<PyFileObject*>(o);
            h = get_file_handle(fo);
        }
        else {
            PyErr_SetString(PyExc_TypeError, "unrecognized I/O object type");
            return false;
        }
    }
    if(! ::CancelIo(h)) {
        PyErr_SetString(PyExc_RuntimeError, "::CancelIo failed");
        return false;
    }
    return true;
}

bool apoll::read(PyFileObject *fo, unsigned long long offset, unsigned long size
                    , ::PyObject *acto)
{
    aop_read * arop = new aop_read(acto, fo, offset, size);
	bool rr = read_impl(arop);
	if(!rr) {
		delete arop;
	}
	return rr;
}

bool apoll::write(PyFileObject *fo, unsigned long long offset
                     , ::PyObject *datao, ::PyObject *acto)
{
    aop_write * awop = new aop_write(acto, fo, offset, datao);
	bool wr = write_impl(awop);
	if(!wr) {
		delete awop;
	}
	return wr;
}

::PyObject * apoll::poll(long ms)
{
    DWORD bytes_transferred = 0;
    ULONG_PTR completion_key = 0;
    aop_root *ovr = 0;
    PyObject * result = PyList_New(0);

    for(;;) {
        BOOL success = FALSE;
        OVERLAPPED *ovl = NULL;
        Py_BEGIN_ALLOW_THREADS;
        success = ::GetQueuedCompletionStatus(iocp_handle_  // CompletionPort
            , &bytes_transferred                                // lpNumberOfBytesTransferred
            , &completion_key                                   // lpCompletionKey
            , &ovl                                              // lpOverlapped
            , ms                                                // dwMilliseconds
            );
        ms = 0; // only first request should wait, so timeout is zero for subsequent ops
        Py_END_ALLOW_THREADS;
        ovr = static_cast<aop_root*>(ovl);

        if ( (0 == ovl) && (FALSE == success) ) {
            if (WAIT_TIMEOUT != ::GetLastError()) {
                PyErr_SetString(PyExc_RuntimeError, "::GetQueuedCompletionStatus failed");
                return NULL;
            }
            else {
                return result;
            }
        }

        if ( 0 == ovl) {
            PyErr_SetString(PyExc_RuntimeError
                , "something impossible happened in ::GetQueuedCompletionStatus");
            return NULL;
        }

        PyObject *op_result = ovr->to_python(success, bytes_transferred);
        PyList_Append(result, op_result);
        Py_XDECREF(op_result);
        delete ovr;
    }

    return result;
}

static PyMethodDef apoll_methods[] = {
    { "accept"
        , reinterpret_cast<PyCFunction>(&apoll::accept_meth)
        , METH_VARARGS
        , "start asynchronous accept operation"
    }
    , { "connect"
        , reinterpret_cast<PyCFunction>(&apoll::connect_meth)
        , METH_VARARGS
        , "start asynchronou connect operation"
    }
    , { "send"
        , reinterpret_cast<PyCFunction>(&apoll::send_meth)
        , METH_VARARGS
        , "start asynchronous send operation"
    }
    , { "sendto"
        , reinterpret_cast<PyCFunction>(&apoll::sendto_meth)
        , METH_VARARGS
        , "start asynchronous sendto operation"
    }
    , { "recv"
        , reinterpret_cast<PyCFunction>(&apoll::recv_meth)
        , METH_VARARGS
        , "start asynchronous recv operation"
    }
    , { "recvfrom"
        , reinterpret_cast<PyCFunction>(&apoll::recvfrom_meth)
        , METH_VARARGS
        , "start asynchronous recvfrom operation"
    }
    , { "poll"
        , reinterpret_cast<PyCFunction>(&apoll::poll_meth)
        , METH_VARARGS
        , "poll asynchronous operation results"
    }
    , { "cancel"
        , reinterpret_cast<PyCFunction>(&apoll::cancel_meth)
        , METH_VARARGS
        , "cancel asynchronous operations for given object"
    }
    , { "read"
        , reinterpret_cast<PyCFunction>(&apoll::read_meth)
        , METH_VARARGS
        , "start asynchronous read operation on file"
    }
    , { "write"
        , reinterpret_cast<PyCFunction>(&apoll::write_meth)
        , METH_VARARGS
        , "start asynchronous write operation on file" 
    }
    , {NULL} /* Sentinel */
};

PyTypeObject apoll_Type = {
    PyObject_HEAD_INIT(NULL)
    0,                         /*ob_size*/
    "_pyasynchio.apoll",             /*tp_name*/
    sizeof(apoll), /*tp_basicsize*/
    0,                         /*tp_itemsize*/
    (destructor)apoll::dealloc,                         /*tp_dealloc*/
    0,                         /*tp_print*/
    0,                         /*tp_getattr*/
    0,                         /*tp_setattr*/
    0,                         /*tp_compare*/
    0,                         /*tp_repr*/
    0,                         /*tp_as_number*/
    0,                         /*tp_as_sequence*/
    0,                         /*tp_as_mapping*/
    0,                         /*tp_hash */
    0,                         /*tp_call*/
    0,                         /*tp_str*/
    0,                         /*tp_getattro*/
    0,                         /*tp_setattro*/
    0,                         /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,        /*tp_flags*/
    "apoll objects",           /* tp_doc */

    0,                                  // tp_traverse 
    0,                                  // tp_clear
    0,                                  // tp_richcompare 
    0,                                  // tp_weaklistoffset 

    0,                                  // tp_iter
    0,                                  // tp_iternext 

    apoll_methods,                      // tp_methods 
    0,                                  // tp_members
    0,                                  // tp_getset
    0,                                  // tp_base 
    
    0,                                  // tp_dict
    0,                                  // tp_descr_get
    0,                                  // tp_descr_set
    0,                                  // tp_dictoffset 
    apoll::init,               // tp_init
    0,                                  // tp_alloc 
    PyType_GenericNew,                 // tp_new
    0,                                  // tp_free
    0,                                  // tp_is_gc
    0,                                  // tp_bases
    0,                                  // tp_mro
    0,                                  // tp_cache
    0,                                  // tp_subclasses
    0,                                  // tp_weaklist
    0,                                  // tp_del
};


int apoll::init(PyObject *self, PyObject *args, PyObject *kwds)
{
    try {
        new (self) pyasynchio::apoll(args, kwds);
    }
    catch(std::exception &)
    {
        return FALSE;
    }
    return TRUE;
}

void apoll::dealloc(pyasynchio::apoll *self)
{
    self->~apoll();
    self->ob_type->tp_free(self);
}

PyObject* apoll::accept_meth(apoll *self, ::PyObject *args)
{
    PyObject *listen_sock_raw, *accept_sock_raw
        , *lsock_ref_obj, *asock_ref_obj
        , *act;
    if (!PyArg_ParseTuple(args, "OOOOO:accept", &listen_sock_raw, &accept_sock_raw
        , &lsock_ref_obj, &asock_ref_obj
        , &act)) {
        return NULL;
    }

    PySocketSockObject *listen_sock = py_convert<PySocketSockObject>(listen_sock_raw
        , socketmodule_api.Sock_Type);
    if (NULL == listen_sock) {
        return NULL;
    }

    PySocketSockObject *accept_sock = py_convert<PySocketSockObject>(accept_sock_raw
        , socketmodule_api.Sock_Type);
    if (NULL == accept_sock) {
        return NULL;
    }

    if(!self->accept(listen_sock, accept_sock, lsock_ref_obj, asock_ref_obj, act)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::connect_meth(apoll *self, ::PyObject *args)
{
    PyObject *so_raw, *so_ref, *addro, *acto;
    if (!PyArg_ParseTuple(args, "OOOO:connect", &so_raw, &so_ref, &addro, &acto)) {
        return NULL;
    }

    PySocketSockObject *so = py_convert<PySocketSockObject>(so_raw
        , socketmodule_api.Sock_Type);
    if (NULL == so) {
        return NULL;
    }

    if (!self->connect(so, so_ref, addro, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::sendto_meth(apoll * self, ::PyObject * args)
{
    ::PyObject *so_raw, *so_ref, *addro, *datao, *acto;
    unsigned long flags;
    if (!PyArg_ParseTuple(args, "OOOOkO:sendto", &so_raw, &so_ref, &addro, &datao, &flags, &acto)) {
        return NULL;
    }

    PySocketSockObject *so = py_convert<PySocketSockObject>(so_raw
        , socketmodule_api.Sock_Type);
    if (NULL == so) {
        return NULL;
    }

    if (!self->sendto(so, so_ref, addro, datao, flags, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::send_meth(apoll * self, ::PyObject * args)
{
    ::PyObject *so_raw, *so_ref, *datao, *acto;
    unsigned long flags;
    if (!PyArg_ParseTuple(args, "OOOkO:send", &so_raw, &so_ref
        , &datao, &flags, &acto)) {
        return NULL;
    }

    PySocketSockObject *so = py_convert<PySocketSockObject>(so_raw
        , socketmodule_api.Sock_Type);
    if (NULL == so) {
        return NULL;
    }

    if (!self->send(so, so_ref, datao, flags, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::recv_meth(apoll *self, ::PyObject *args)
{
    ::PyObject *so_raw, *so_ref, *acto;
    unsigned long flags, size;
    if (!PyArg_ParseTuple(args, "OOkkO:recv", &so_raw, &so_ref
        , &size, &flags, &acto)) {
        return NULL;
    }

    PySocketSockObject *so = py_convert<PySocketSockObject>(so_raw
        , socketmodule_api.Sock_Type);
    if (NULL == so) {
        return NULL;
    }

    if (!self->recv(so, so_ref, size, flags, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::recvfrom_meth(apoll *self, ::PyObject *args)
{
    ::PyObject *so_raw, *so_ref, *acto;
    unsigned long flags, size;
    if (!PyArg_ParseTuple(args, "OOkkO:recvfrom", &so_raw, &so_ref
        , &size, &flags, &acto)) {
        return NULL;
    }

    PySocketSockObject *so = py_convert<PySocketSockObject>(so_raw
        , socketmodule_api.Sock_Type);
    if (NULL == so) {
        return NULL;
    }

    if (!self->recvfrom(so, so_ref, size, flags, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::poll_meth(apoll *self, ::PyObject *args)
{
    long ms;
    if (!PyArg_ParseTuple(args, "l:poll", &ms)) {
        return NULL;
    }

    return self->poll(ms);
}

::PyObject * apoll::cancel_meth(apoll * self, ::PyObject * args)
{
    ::PyObject * o;
    if (!PyArg_ParseTuple(args, "O:cancel", &o)) {
        return NULL;
    }

    if (!self->cancel(o)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::write_meth(apoll * self, ::PyObject * args)
{
    ::PyObject * fo_raw, *acto, *datao;
    unsigned long long offset = 0;
    if(!PyArg_ParseTuple(args, "OKOO:write", &fo_raw, &offset, &datao, &acto)) {
        return NULL;
    }

    PyFileObject * fo = py_convert<PyFileObject>(fo_raw, &PyFile_Type);
    if(NULL == fo) {
        return NULL;
    }

    if(!self->write(fo, offset, datao, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

::PyObject * apoll::read_meth(apoll * self, ::PyObject * args)
{
    ::PyObject * fo_raw, *acto;
    unsigned long long offset = 0;
    unsigned long size = 0;
    if (!PyArg_ParseTuple(args, "OKkO:read", &fo_raw, &offset, &size, &acto)) {
        return NULL;
    }

    PyFileObject * fo = py_convert<PyFileObject>(fo_raw, &PyFile_Type);
    if (NULL == fo) {
        return NULL;
    }

    if (!self->read(fo, offset, size, acto)) {
        return NULL;
    }

    Py_INCREF(Py_None);
    return Py_None;
}

} // namespace pyasynchio